test/cram/pattern.t

```diff
@@ -37,54 +120,119 @@ Selecting both misc::passing and misc::failing
+
+  $ simple.c.bin --pattern 'misc/*!(passing)' --verbose
+  [\x1b[0;34m----\x1b[0m] Criterion v2.2.1 (esc)
+  [\x1b[0;34m====\x1b[0m] Running \x1b[0;34m2\x1b[0m tests from \x1b[0;33mmisc\x1b[0m: (esc)
+  [\x1b[0;34mRUN \x1b[0m] misc::failing (esc)
+  [\x1b[0;34m----\x1b[0m] \x1b[0;1msimple.c\x1b[0m:\x1b[0;31m4\x1b[0m: Assertion failed: The expression 0 is false. (esc)
+  [\x1b[0;31mFAIL\x1b[0m] misc::failing (esc)
+  [\x1b[0;34mRUN \x1b[0m] misc::passing (esc)
+  [\x1b[0;32mPASS\x1b[0m] misc::passing (esc)
+  [\x1b[0;34m====\x1b[0m] \x1b[0;1mSynthesis: Tested: \x1b[0;34m2\x1b[0;1m | Passing: \x1b[0;32m1\x1b[0;1m | Failing: \x1b[0;31m1\x1b[0;1m | Crashing: \x1b[0m0\x1b[0;1m \x1b[0m (esc)
+
+
+
+
+
 Testing nested extglob patterns
 
-  $ simple.c.bin --pattern '@(+(nest)ed))' --verbose
+  $ simple.c.bin --pattern '@(+(nest)ed)' --verbose
   [\x1b[0;34m----\x1b[0m] Criterion v2.2.1 (esc)
   [\x1b[0;34m====\x1b[0m] Running \x1b[0;34m2\x1b[0m tests from \x1b[0;33mmisc\x1b[0m: (esc)
   [\x1b[0;33mSKIP\x1b[0m] misc::passing: Test is disabled (esc)
   [\x1b[0;33mSKIP\x1b[0m] misc::failing: Test is disabled (esc)
   [\x1b[0;34m====\x1b[0m] \x1b[0;1mSynthesis: Tested: \x1b[0;34m0\x1b[0;1m | Passing: \x1b[0;32m0\x1b[0;1m | Failing: \x1b[0m0\x1b[0;1m | Crashing: \x1b[0m0\x1b[0;1m \x1b[0m (esc)
 
-Testing one or more
-
   $ simple.c.bin --pattern '?(*(a|b))' --verbose
   [\x1b[0;34m----\x1b[0m] Criterion v2.2.1 (esc)
   [\x1b[0;34m====\x1b[0m] Running \x1b[0;34m2\x1b[0m tests from \x1b[0;33mmisc\x1b[0m: (esc)
   [\x1b[0;33mSKIP\x1b[0m] misc::passing: Test is disabled (esc)
   [\x1b[0;33mSKIP\x1b[0m] misc::failing: Test is disabled (esc)
   [\x1b[0;34m====\x1b[0m] \x1b[0;1mSynthesis: Tested: \x1b[0;34m0\x1b[0;1m | Passing: \x1b[0;32m0\x1b[0;1m | Failing: \x1b[0m0\x1b[0;1m | Crashing: \x1b[0m0\x1b[0;1m \x1b[0m (esc)
 
+  $ simple.c.bin --pattern 'misc/pa@(s|*(s))ing' --verbose
+  [\x1b[0;34m----\x1b[0m] Criterion v2.2.1 (esc)
+  [\x1b[0;34m====\x1b[0m] Running \x1b[0;34m2\x1b[0m tests from \x1b[0;33mmisc\x1b[0m: (esc)
+  [\x1b[0;34mRUN \x1b[0m] misc::passing (esc)
+  [\x1b[0;32mPASS\x1b[0m] misc::passing (esc)
+  [\x1b[0;33mSKIP\x1b[0m] misc::failing: Test is disabled (esc)
+  [\x1b[0;34m====\x1b[0m] \x1b[0;1mSynthesis: Tested: \x1b[0;34m1\x1b[0;1m | Passing: \x1b[0;32m1\x1b[0;1m | Failing: \x1b[0m0\x1b[0;1m | Crashing: \x1b[0m0\x1b[0;1m \x1b[0m (esc)
+
+
+
+
 Testing malformed pattern
 
   $ simple.c.bin --pattern '?(malformed' --verbose
-  pattern error: mismatching parenthesis
-  [1]
+  pattern error: Mismatching parenthesis
+  [3]
 
-Testing range negation
+  $ simple.c.bin --pattern '(malformed' --verbose
+  pattern error: Unexpected character '('
+  [3]
 
-  $ simple.c.bin --pattern 'misc/[!azerty]assing' --verbose
+  $ simple.c.bin --pattern 'misc/passing)' --verbose
+  pattern error: Unexpected character ')'
+  [3]
+
+  $ simple.c.bin --pattern 'misc/[' --verbose
```

Comment:
I think that mismatching square brackets should also yield an error message, for the same reasons as the `[p-f]` pattern. We don't need to closely follow the spec for these cases, as we work with C identifiers rather than filenames.